File: itx-java/itx-impl/src/main/java/io/github/chehsunliu/itx/impl/repo/mariadb/MariaDbUserRepo.java
package io.github.chehsunliu.itx.impl.repo.mariadb;

import static io.github.chehsunliu.itx.impl.repo.Jdbc.bindAll;
import static io.github.chehsunliu.itx.impl.repo.Jdbc.firstOrNull;
import static io.github.chehsunliu.itx.impl.repo.Jdbc.useConnection;

import io.github.chehsunliu.itx.contract.repo.RepoNotFoundException;
import io.github.chehsunliu.itx.contract.repo.User;
import io.github.chehsunliu.itx.contract.repo.UserRepo;
import java.util.UUID;
import javax.sql.DataSource;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
final class MariaDbUserRepo implements UserRepo {
    private final DataSource ds;

    @Override
    public User upsert(UpsertParams params) {
        return useConnection(ds, conn -> {
            try (var ps = conn.prepareStatement(
                    "INSERT INTO users (id, email) VALUES (?, ?) ON DUPLICATE KEY UPDATE id = id")) {
                bindAll(ps, params.id().toString(), params.email());
                ps.executeUpdate();
            }
            try (var ps = conn.prepareStatement("SELECT id, email FROM users WHERE id = ?")) {
                bindAll(ps, params.id().toString());
                try (var rs = ps.executeQuery()) {
                    return firstOrNull(rs, r -> new User(UUID.fromString(r.getString("id")), r.getString("email")))
                            .orElseThrow(() -> new IllegalStateException("upsert returned no row"));
                }
            }
        });
    }

    @Override
    public User get(UUID id) {
        return useConnection(ds, conn -> {
            try (var ps = conn.prepareStatement("SELECT id, email FROM users WHERE id = ?")) {
                bindAll(ps, id.toString());
                try (var rs = ps.executeQuery()) {
                    return firstOrNull(rs, r -> new User(UUID.fromString(r.getString("id")), r.getString("email")))
                            .orElseThrow(RepoNotFoundException::new);
                }
            }
        });
    }
}

File: itx-kt/itx-impl/src/main/kotlin/io/github/chehsunliu/itx/impl/repo/mariadb/MariaDbUserRepo.kt
package io.github.chehsunliu.itx.impl.repo.mariadb

import io.github.chehsunliu.itx.contract.repo.RepoNotFoundException
import io.github.chehsunliu.itx.contract.repo.User
import io.github.chehsunliu.itx.contract.repo.UserRepo
import io.github.chehsunliu.itx.impl.repo.bindAll
import io.github.chehsunliu.itx.impl.repo.firstOrNull
import io.github.chehsunliu.itx.impl.repo.useConnection
import java.util.UUID
import javax.sql.DataSource

internal class MariaDbUserRepo(
    private val ds: DataSource,
) : UserRepo {
    override suspend fun upsert(params: UserRepo.UpsertParams): User =
        ds.useConnection { conn ->
            conn
                .prepareStatement(
                    "INSERT INTO users (id, email) VALUES (?, ?) ON DUPLICATE KEY UPDATE id = id",
                ).use {
                    it.bindAll(params.id.toString(), params.email)
                    it.executeUpdate()
                }
            conn.prepareStatement("SELECT id, email FROM users WHERE id = ?").use { ps ->
                ps.bindAll(params.id.toString())
                ps.executeQuery().use { rs ->
                    rs.firstOrNull { User(UUID.fromString(it.getString("id")), it.getString("email")) }
                        ?: error("upsert returned no row")
                }
            }
        }

    override suspend fun get(id: UUID): User =
        ds.useConnection { conn ->
            conn.prepareStatement("SELECT id, email FROM users WHERE id = ?").use { ps ->
                ps.bindAll(id.toString())
                ps.executeQuery().use { rs ->
                    rs.firstOrNull { User(UUID.fromString(it.getString("id")), it.getString("email")) }
                        ?: throw RepoNotFoundException()
                }
            }
        }
}
